### PR TITLE
Add JSON format to graph command

### PR DESCRIPTION
**Pull request by neakor**
This allows the `graph` command to export the project graph into JSON format. Users can then parse the JSON data and perform operations such as changeset detection, detecting which Xcode project is affected if a dependency is affected.

For example, given a PR, we can detect which Swift packages are affected by the files changed in the PR. Assuming the repository contains multiple Xcode projects, we then need to figure out which Xcode project is affected by the changed Swift packages. Finally we can then run tests on these affected Xcode projects.

Resolves N/A
Request for comments document (if applies): N/A

### Short description 📝

RFC https://github.com/tuist/tuist/discussions/3629

This allows the `graph` command to export the project graph into JSON format. Users can then parse the JSON data and perform operations such as changeset detection, detecting which Xcode project is affected if a dependency is affected.

### Checklist ✅

- [✅] The code architecture and patterns are consistent with the rest of the codebase.
- [✅] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [✅] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [✅] In case the PR introduces changes that affect users, the documentation has been updated.
- [✅] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

**Review by fortmarek — COMMENTED** [review]
Hey, @neakor!

First off, thanks a lot for the contribution!

> For example, given a PR, we can detect which Swift packages are affected by the files changed in the PR. Assuming the repository contains multiple Xcode projects, we then need to figure out which Xcode project is affected by the changed Swift packages. Finally we can then run tests on these affected Xcode projects.

Could you elaborate a little bit for which changes are you looking for? In the new [Plugins 2.0](https://github.com/tuist/tuist/discussions/3491) re-architecture, we wanted to use this command to output a graph as well. However, encoding `TuistGraph.Graph` might not be the best solution for user-facing data since we don't make any assurances that the API for `TuistGraph.Graph` will stay stable. Plus users have to deal with additional complexity of the project that might not be necessary since some things are only useful for internal usage.

My suggestion for this command is to create a new graph model that we could eventually make stable. It would also be written in such a way, so it's easy for people to use. Then for tuist plugins, we could reuse `ProjectAutomation` product that would contain this model and you could use it in Swift executable code, so it's very easy to work with.

So, for example, you could achieve the following API:

```
// MyPlugin.swift executable file

import ProjectAutomation

let graph = TuistGraphLoader().loadGraph()
graph.targets.flatMap(\.sources) {
    // do something with all the sources
}
```

What do you think, @neakor? cc @luispadron @kwridan since we have discussed this on the last tuist get together.

**Comment by luispadron** [comment]
@fortmarek I agree with your comment about using a new model. Exposing the internal graph model may lead to breaking changes in non-major versions of tuist which wouldn't be great! I think a new model where we can slowly add new properties to when things are stable is the appropriate move here. 

In terms of how to get this, i think a command in tuist to dump this is great! Also, being able to use it in tasks is also important so it needs to live somewhere shareable.

**Comment by neakor** [comment]
@fortmarek @luispadron These are great suggestions. I can certainly create a new model and export that one instead. Do you guys have any opinions on what the model would look like? For our use case, we only need the `packages` property. I can just simply make a direct copy of the existing `Graph` model for now.

**Comment by danyf90** [comment]
What do you actually need? Isn't it already part of the `tuist dump project` output?

Also, for changes like this, I think it would be better to pass though an RFC before doing the actual implementation 🙏

**Comment by fortmarek** [comment]
> Also, for changes like this, I think it would be better to pass though an RFC before doing the actual implementation 🙏

Good point, I agree with this.

> I can certainly create a new model and export that one instead. Do you guys have any opinions on what the model would look like?

Something similar to `TuistGraph.Graph` but simpler and with primitive types (e.g. using `String` instead of `AbsolutePath`):

```swift
struct Graph {
   // Map of name of target to `Target`
   let targets: [String: Target]
   ...
}

struct Target {
   let name: String
   ...
}
```

It's imho a matter of going through what the current `Graph` has and then extracting the useful bits. If you wanted to make an RFC for this, that'd be great! I'm also up for chatting in our [Slack](https://join.slack.com/t/tuistapp/shared_invite/zt-g38gajhj-D6LLakrPnVCy4sLm24KxaQ).

You also already might be able to use `tuist dump` but `tuist graph --json` should be easy to work with (and once we have a special product for it, you'll get decoding that for free)

**Comment by neakor** [comment]
> What do you actually need? Isn't it already part of the `tuist dump project` output?
> 
> Also, for changes like this, I think it would be better to pass though an RFC before doing the actual implementation 🙏

We need the full graph output including the path. Our requirement is to find the Xcode projects that include a set of Swift packages. The `dump` command does not provide enough information for us to make that mapping. Specifically it lacks the actual paths of the Swift packages in the output JSON.

We are merely adding a new format in addition to the two existing ones, `png` and `dot`. I feel like an RFC is too heavy handed isn't it?

**Comment by neakor** [comment]
@fortmarek regarding the new struct, that sounds good! I can update this PR if you guys are onboard with the general direction. Please let me know.

**Comment by danyf90** [comment]
> We are merely adding a new format in addition to the two existing ones, png and dot. I feel like an RFC is too heavy handed isn't it?

There are a lot of different way to achieve the same result and discussing it before implementing (even on Slack if an RFC seems too much) is always a good idea.

For example, another possible approach could have been to add a different kind to `tuist dump`, e.g. `tuist dump graph`, as `tuist dump` is the command responsible to print structures in JSON format.

Considering all the possible approaches, I think adding the json type to `tuist graph` (as you did) is the best one for now.
I agree with the others that we should make the encoding independent from the `TuistGraph`, so that we can evolve that independently from the output format

**Comment by neakor** [comment]
@danyf90 they are really valid points! Sorry I didn't know about the Slack channel and discussions.

How should we proceed now? Should I just update this PR with the separate model or do we need to go through some other process?

**Comment by fortmarek** [comment]
> How should we proceed now? Should I just update this PR with the separate model or do we need to go through some other process?

I think the best way forward is to create an RFC as @danyf90 suggests. You are welcome to join the Slack channel for more lightweight discussion, though! The RFC does not have to be heavy-handed - in this case, it should primarily include how the model should look like and whether we want it to be a part of `tuist graph --json` (which I think we should)

I'm up for keeping this PR opened but would wait with implementation changes until we align ourselves in the RFC.

**Comment by neakor** [comment]
That sounds good! Thank you for the suggestions. I'll create an RFC for this.

**Comment by fortmarek** [comment]
@neakor sure! We don't have a rigorous process - just create a new topic [here](https://github.com/tuist/tuist/discussions/categories/rfcs), detail some of the background why you want to make the change and then add some details how you actually want to achieve the final result.

**Comment by neakor** [comment]
@fortmarek thank you for the tip!

Everyone, here's the RFC https://github.com/tuist/tuist/discussions/3629

File: Sources/TuistKit/Services/GraphService.swift [hunk 1]
```diff
@@ -11,6 +11,10 @@ import TuistSupport
 final class GraphService {
     private let graphVizMapper: GraphToGraphVizMapping
     private let manifestGraphLoader: ManifestGraphLoading
+    
+    private enum GraphServiceError: Error {
```
Comment (pepibumur) [inline Sources/TuistKit/Services/GraphService.swift:15]: To be consistent with the rest of the codebase, I'd recommend to conform the `FatalError` protocol, and use cases to represent the error type. `FatalError` requires errors to include a description, which is presented automatically by an error handler that catches fatal errors when they bubble up, and the type is used to determine whether the error should get automatically reported.


```swift
private enum GraphServiceError: FatalError {
  case invalidFormat(String)

  var description: String {
    switch self {
      case .invalidFormat(let format): "....."
    }
  }

  var type: ErrorType {
    switch self {
      case .invalidFormat: return .abort
    }
  }
}
```

File: Sources/TuistKit/Services/GraphService.swift [hunk 2]
```diff
@@ -67,8 +84,14 @@ final class GraphService {
             try exportDOTRepresentation(from: graph, at: filePath)
         case .png:
             try exportPNGRepresentation(from: graph, at: filePath, layoutAlgorithm: layoutAlgorithm)
+        default:
+            throw GraphServiceError.withMessage("\(format.rawValue) is not a visual graph format to export")
         }
     }
+    
+    private func export(jsonContent: String, at filePath: AbsolutePath) throws {
+        try FileHandler.shared.write(jsonContent, path: filePath, atomically: true)
```
Comment (pepibumur) [inline Sources/TuistKit/Services/GraphService.swift:93]: Because the method is simple and it's only used called from a place, I recommend inlining the code here in the `run` method.

File: Tests/TuistKitTests/Services/GraphServiceTests.swift [hunk 3]
```diff
@@ -64,4 +64,33 @@ final class GraphServiceTests: TuistUnitTestCase {
...
+            format: .json,
+            layoutAlgorithm: .dot,
+            skipTestTargets: false,
+            skipExternalDependencies: false,
+            targetsToFilter: [],
+            path: temporaryPath,
+            outputPath: temporaryPath
+        )
+        let got = try FileHandler.shared.readTextFile(graphPath)
+        let expected = "{\"packages\":[],\"workspace\":{\"path\":\"\\/\",\"schemes\":[],\"additionalFiles\":[],\"name\":\"test\",\"projects\":[],\"xcWorkspacePath\":\"\\/\"},\"targets\":[],\"path\":\"\\/\",\"dependencies\":[],\"name\":\"graph\",\"projects\":[]}"
```
Comment (pepibumur) [inline Tests/TuistKitTests/Services/GraphServiceTests.swift:88]: Comparing the raw JSON string will make the test cumbersome to maintain because changes in the graph will lead to changes in the string and thus failing tests. What about serializing that string back into a graph model?

File: projects/docs/docs/commands/graph.md [hunk 4]
```diff
@@ -15,6 +15,7 @@ saying goes, "one image is worth a thousand words":
 The command will output the dependency graph as an image, in the `png` format.
 You can also change the format to `dot` (see [DOT](<https://en.wikipedia.org/wiki/DOT_(graph_description_language)>))
 to get the raw contents of the graph.
+You can also change the format to `json` to export the project structure into a JSON file.
```
Comment (pepibumur) [inline projects/docs/docs/commands/graph.md:18]: Because we are exporting a JSON, I anticipate users building workflows upon that. That means the output becomes a public interface between the users and the CLI. Until the Graph's structure stabilizes, I'd add an [admonition](https://docusaurus.io/docs/markdown-features/admonitions) here making it clear that the format is subject to breaking changes.

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 - Schemes can be hidden from the dropdown menu `Scheme(hidden: true)` [#3598](https://github.com/tuist/tuist/pull/3598) by [@pepibumur](https://github.com/pepibumur)
 - Sort schemes alphabetically by default [#3598](https://github.com/tuist/tuist/pull/3598) by [@pepibumur](https://github.com/pepibumur)
 - Add automation to release [#3603](https://github.com/tuist/tuist/pull/3603/) by [@luispadron](https://github.com/luispadron)
+- JSON format to `graph` command to support exporting a project's structure to a JSON file.
```
Comment (pepibumur) [inline CHANGELOG.md:16]: @neakor don't forget to add the link to your user and PR so that we attribute the improvement to you.

**Review by pepibumur — APPROVED** [review]
Thanks a lot for landing this contribution @neakor 👏🏼. You nailed it!

**Comment by pepibumur** [comment]
@all-contributors add @neakor for code

**Comment by pepibumur** [comment]
@neakor I added you as contributor so that you no longer have to make contributions too forks, which are very inconvenient.

**Comment by neakor** [comment]
@pepibumur thank you for the review! As suggested by other contributors, I still need to update this PR to create the new graph model after the RFC is approved. I'm just waiting on that.

File: CHANGELOG.md (hunk 5 above)
Comment (neakor) [reply CHANGELOG.md:16]: Sounds good!

File: Sources/TuistKit/Services/GraphService.swift (hunk 1 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:15]: Thank you for the suggestion. Let's do that.

File: Sources/TuistKit/Services/GraphService.swift (hunk 2 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:93]: Will do!

File: Tests/TuistKitTests/Services/GraphServiceTests.swift (hunk 3 above)
Comment (neakor) [reply Tests/TuistKitTests/Services/GraphServiceTests.swift:88]: Yeah sure. I thought comparing the direct output would be more accurate. But I suppose deserialize it again would work too.

File: projects/docs/docs/commands/graph.md (hunk 4 above)
Comment (neakor) [reply projects/docs/docs/commands/graph.md:18]: Sounds good!

**Comment by neakor** [comment]
@pepibumur I've updated the PR based on your feedback. Please take another look.

**Comment by neakor** [comment]
@fortmarek can you please review as well? I think we need two reviewers.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 6]
```diff
@@ -0,0 +1,142 @@
+import Foundation
+import TSCBasic
+
+public struct GraphOutputJSON: Codable, Equatable {
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:4]: I like the name `GraphOutput` a bit more. I don't think the `JSON` suffix is necessary here.

Comment (fortmarek) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:4]: these are all public methods as well, so I think we should document the models and public properties.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    public var projects: [String: ProjectOutputJSON]
+    
+    public init(name: String, path: String, projects: [String: ProjectOutputJSON]) {
+        self.name = name
+        self.path = path
+        self.projects = projects
+    }
+    
+    public static func from(_ graph: Graph, withConfig config: Config) -> GraphOutputJSON {
+        var projects = [String: ProjectOutputJSON]()
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:16]: I think we should create methods for the individual models, such as:
```swift
public static func from(_ project: Project, with config: Config) -> ProjectOutput {...}
```

instead of putting everything into a single function.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        
+        return GraphOutputJSON(name: graph.name, path: graph.path.pathString, projects: projects)
+    }
+}
+
+public struct ProjectOutputJSON: Codable, Equatable {
+    public var name: String
+    public var path: String
+    public var packages: [PackageOutputJSON]?
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:80]: do we want this to be optional? It might be confusing for users, so it might be worth having just an empty array instead if there are no packages (I'd do the same for `targets` and `schemes` as well)

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                projectJSON.targets = targets
+            }
+            projects[project.path.pathString] = projectJSON
+        }
+        
+        return GraphOutputJSON(name: graph.name, path: graph.path.pathString, projects: projects)
+    }
+}
+
+public struct ProjectOutputJSON: Codable, Equatable {
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:77]: I'd create a new file for this one.

While you do that, I'd also create a new folder in `TuistGraph` such as `TuistGraph/Output`. Once we decide to extract the models into their own product, we can do that very easily by just moving all the files in that directory

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            
+            // Convert targets.
+            let didDisableAutogeneratedSchemes = config.disableAutogeneratedSchemes
+            var targets = [TargetOutputJSON]()
+            for target in project.targets {
+                targets.append(TargetOutputJSON(name: target.name, product: target.product.rawValue))
+                
+                // For each target, a default scheme is generated. Include these in the schemes output.
+                // If the target's product is a testing one, include the target as test action.
+                if !didDisableAutogeneratedSchemes {
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:50]: I'm not sure if I completely follow this one. Should not autogenerated schemes be part of the graph already? Ideally, this `from` function should not contain a lot of logic like this as it would be then coupled to logic in mappers which will make it easy to make mistakes.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                        testTargets.append(testTarget.target.name)
+                    }
+                }
+                schemes.append(SchemeOutputJSON(name: scheme.name, testActionTargets: testTargets))
+            }
+            
+            // Convert targets.
+            let didDisableAutogeneratedSchemes = config.disableAutogeneratedSchemes
+            var targets = [TargetOutputJSON]()
+            for target in project.targets {
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:45]: If you have specific `from` methods for a target, you could do something along the lines of:
```swift
let targets = project.targets.reduce(into: []) {
   $0.append(TargetOutput.from($1))
}
```

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift [hunk 12]
```diff
@@ -0,0 +1,142 @@
+import Foundation
+import TSCBasic
+
+public struct GraphOutputJSON: Codable, Equatable {
+    public var name: String
```
Comment (fortmarek) [inline Sources/TuistGraph/Graph/GraphOutputJSON.swift:5]: is there a specific reason for having all the properties as variables? I'm not sure that's completely necessary.

File: Sources/TuistKit/Services/GraphService.swift [hunk 13]
```diff
@@ -67,6 +83,8 @@ final class GraphService {
             try exportDOTRepresentation(from: graph, at: filePath)
         case .png:
             try exportPNGRepresentation(from: graph, at: filePath, layoutAlgorithm: layoutAlgorithm)
+        default:
```
Comment (fortmarek) [inline Sources/TuistKit/Services/GraphService.swift:86]:
```suggestion
        case .json:
```

File: Sources/TuistKit/Services/GraphService.swift [hunk 14]
```diff
@@ -98,3 +116,39 @@ final class GraphService {
...
+            return "\(format) is not valid for export"
+        case .encodingError(let format):
+            return "failed to encode graph to \(format)"
+        }
+    }
+    
+    var type: ErrorType {
+        switch self {
+        case .invalidFormat:
+            return .abort
```
Comment (fortmarek) [inline Sources/TuistKit/Services/GraphService.swift:136]: I think this should be a `.bug` since it should never happen, right?

File: Tests/TuistKitTests/Services/GraphServiceTests.swift [hunk 15]
```diff
@@ -64,4 +65,34 @@ final class GraphServiceTests: TuistUnitTestCase {
...
+            layoutAlgorithm: .dot,
+            skipTestTargets: false,
+            skipExternalDependencies: false,
+            targetsToFilter: [],
+            path: temporaryPath,
+            outputPath: temporaryPath
+        )
+        let got = try FileHandler.shared.readTextFile(graphPath)
+        
+        let result = try! JSONDecoder().decode(GraphOutputJSON.self, from: got.data(using: .utf8)!)
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/GraphServiceTests.swift:90]:
```suggestion
        let result = try JSONDecoder().decode(GraphOutputJSON.self, from: got.data(using: .utf8)!)
```

File: Tests/TuistKitTests/Services/GraphServiceTests.swift [hunk 16]
```diff
@@ -64,4 +65,34 @@ final class GraphServiceTests: TuistUnitTestCase {
...
+        // Then
+        XCTAssertEqual(result, GraphOutputJSON.init(name: "graph", path: "/", projects: [String: ProjectOutputJSON]()))
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/GraphServiceTests.swift:92]:
```suggestion
        XCTAssertEqual(result, GraphOutputJSON(name: "graph", path: "/", projects: [:]))
```

File: Sources/TuistKit/Services/GraphService.swift [hunk 17]
```diff
@@ -41,19 +47,29 @@ final class GraphService {
...
+                skipExternalDependencies: skipExternalDependencies,
+                targetsToFilter: targetsToFilter
+            )
+            
+            try export(graph: graphVizGraph, at: filePath, withFormat: format, layoutAlgorithm: layoutAlgorithm)
+        case .json:
+            let config = try configLoader.loadConfig(path: path)
+            
+            let outputGraph = GraphOutputJSON.from(graph, withConfig: config)
+            try outputGraph.export(to: filePath)
```
Comment (fortmarek) [inline Sources/TuistKit/Services/GraphService.swift:70]: I think we should make the `path` optional (don't default to `FileHandler.shared.currentPath` in `GraphCommand`) and output JSON to stdout by default. This will make it easier to reuse the output in a different process.

File: projects/docs/docs/commands/graph.md [hunk 18]
```diff
@@ -15,6 +15,10 @@ saying goes, "one image is worth a thousand words":
 The command will output the dependency graph as an image, in the `png` format.
 You can also change the format to `dot` (see [DOT](<https://en.wikipedia.org/wiki/DOT_(graph_description_language)>))
 to get the raw contents of the graph.
+:::caution
+You can also change the format to `json` to export the project structure into a JSON file. The JSON schema is subject
```
Comment (fortmarek) [inline projects/docs/docs/commands/graph.md:19]:
```suggestion
You can also change the format to `json` to export the graph as a JSON. The JSON schema is subject
```

**Review by fortmarek — CHANGES_REQUESTED** [review]
Hey, @neakor, thanks for the PR! Sorry for the later review but since this is going to be an important part of the project, I think we should take our time implementing this. I added a couple of comments that are worth addressing.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift (hunk 6 above)
Comment (neakor) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:4]: Sounds good.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift (hunk 9 above)
Comment (neakor) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:77]: Sounds good!

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift (hunk 10 above)
Comment (neakor) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:50]: That doesn't seem to be the case actually. When I ran this command on the tuist project, it did not output the schemes.

I didn't want to change the mapping logic since I'd imagine that's used in far more places. There could be a legit reason why they are not in the `Graph` instance.

File: Sources/TuistKit/Services/GraphService.swift (hunk 17 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:70]: That deviates from the other two formats though. Besides, stdout can have a number of outputs in the future. It seems less future-proof and more error-prone to rely on stdout.

Comment (fortmarek) [reply Sources/TuistKit/Services/GraphService.swift:70]: > It seems less future-proof and more error-prone to rely on stdout.

I don't understand this part. The way I see it is that `tuist graph --format json` will be mostly used by other processes using the output and it's therefore better as a default instead of a file.

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift (hunk 10 above)
Comment (neakor) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:50]: I'll just remove this for now then. Let's first get to a point where the output schema matches the current `Graph` type. Later we can change/fix anything if needed.

File: Sources/TuistKit/Services/GraphService.swift (hunk 17 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:70]: If somehow in the future a dependency of the command, say one of the mappers or loaders print to stdout, we would have broken the output. SPM actually had this exact same issue. I'm trying to avoid that mistake here.

Comment (fortmarek) [reply Sources/TuistKit/Services/GraphService.swift:70]: ah, ok, I understand now, you convinced me then 🙂

File: Sources/TuistGraph/Graph/GraphOutputJSON.swift (hunk 12 above)
Comment (neakor) [reply Sources/TuistGraph/Graph/GraphOutputJSON.swift:5]: Good point. We'll make all of them `let`.

File: Sources/TuistKit/Services/GraphService.swift (hunk 14 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:136]: Yeah. sounds good.

File: Sources/TuistKit/Services/GraphService.swift (hunk 17 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:70]: Thank you for the suggestion!

File: Sources/TuistKit/Services/GraphService.swift [hunk 19]
```diff
@@ -98,3 +109,39 @@ final class GraphService {
...
+        case .encodingError:
+            return .abort
+        }
+    }
+}
+
+fileprivate extension GraphOutput {
+
+    func export(to filePath: AbsolutePath) throws {
+        let jsonData = try JSONEncoder().encode(self)
```
Comment (ezraberch) [inline Sources/TuistKit/Services/GraphService.swift:139]: Using `.sortedKeys` will make the output consistent, allowing for string comparison.

I also suggest using `.prettyPrinted` to match your sample in the docs and ` .withoutEscapingSlashes` to remove the unnecessary slashes.

```suggestion
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .prettyPrinted, .withoutEscapingSlashes]
        let jsonData = try encoder.encode(self)
```

**Comment by ezraberch** [comment]
Currently, exporting to `png` opens the file after generating, while exporting to `dot` does not. The `json` export here also does not open the file. Shouldn't this behavior be consistent, one way or the other?

**Comment by luispadron** [comment]
@ezraberch Id be in favor of making the behavior **not open** anything.

**Comment by neakor** [comment]
@ezraberch and @luispadron I'm also in favor of "not open".  Although that seems like a separate PR instead of creeping into this one.

File: Sources/TuistKit/Services/GraphService.swift (hunk 19 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:139]: Sounds good!

File: Sources/TuistKit/Services/GraphService.swift [hunk 20]
```diff
@@ -67,6 +76,8 @@ final class GraphService {
             try exportDOTRepresentation(from: graph, at: filePath)
         case .png:
             try exportPNGRepresentation(from: graph, at: filePath, layoutAlgorithm: layoutAlgorithm)
+        case .json:
+            throw GraphServiceError.bug("\(format.rawValue) is not valid for visual export")
```
Comment (danyf90) [inline Sources/TuistKit/Services/GraphService.swift:80]: Can't we define a proper case `jsonNotValidForVisualExport` instead of having a generic `bug` with specific message?

File: Sources/TuistKit/Services/GraphService.swift [hunk 21]
```diff
@@ -67,6 +76,8 @@ final class GraphService {
             try exportDOTRepresentation(from: graph, at: filePath)
         case .png:
             try exportPNGRepresentation(from: graph, at: filePath, layoutAlgorithm: layoutAlgorithm)
```
Comment (danyf90) [inline Sources/TuistKit/Services/GraphService.swift:78]: Not part of this PR, but if the LayoutAlgorithm is only relevant for png, why isn't it an associated value of it?

File: Tests/TuistKitTests/Services/GraphServiceTests.swift [hunk 22]
```diff
@@ -64,4 +65,34 @@ final class GraphServiceTests: TuistUnitTestCase {
         Graph exported to \(graphPath.pathString)
         """)
     }
+    
+    func test_run_whenJson() throws {
```
Comment (danyf90) [inline Tests/TuistKitTests/Services/GraphServiceTests.swift:69]: Checking the result is very simple here, why don't we do it? It would help to catch future regressions in the export

File: projects/docs/docs/commands/graph.md [hunk 23]
```diff
@@ -15,6 +15,10 @@ saying goes, "one image is worth a thousand words":
 The command will output the dependency graph as an image, in the `png` format.
 You can also change the format to `dot` (see [DOT](<https://en.wikipedia.org/wiki/DOT_(graph_description_language)>))
 to get the raw contents of the graph.
+:::caution
+You can also change the format to `json` to export the graph as a JSON file. The JSON schema is subject
+to change and may be breaking across major versions.
+:::
```
Comment (danyf90) [inline projects/docs/docs/commands/graph.md:21]: Should be `minor` version? Otherwise I don't see the point of the caution as a major is breaking by definition

File: projects/docs/docs/commands/graph.md [hunk 24]
```diff
@@ -52,9 +56,84 @@ If you prefer the old style, without different colors and shapes, pass the `--si
...
+            "packages": [
+                {
+                    "kind": "KIND_OF_PACKAGE_REMOTE_OR_LOCAL",
+                    "path": "PATH_OR_URL_OF_PACKAGE"
+                }
+            ]
+        }
+    }
+}
+```
```
Comment (danyf90) [inline projects/docs/docs/commands/graph.md:139]: Are all the paths relative? I would like the generated graph to be the same regardless of the manifest path (hence I would probably remove the top-level path if not needed)

**Review by danyf90 — CHANGES_REQUESTED** [review]

File: projects/docs/docs/commands/graph.md (hunk 24 above)
Comment (fortmarek) [reply projects/docs/docs/commands/graph.md:139]: > Are all the paths relative? I would like the generated graph to be the same regardless of the manifest path (hence I would probably remove the top-level path if not needed)

I am not sure if I agree with this. If e.g. all source paths are relative, it will actually make it harder to work with.

Comment (danyf90) [reply projects/docs/docs/commands/graph.md:139]: Why would it make it harder to work it? if they a relative to the path when the command is invoked I don't see a problem

Comment (fortmarek) [reply projects/docs/docs/commands/graph.md:139]: The scenario I am thinking of is going through `sources` to e.g. lint all the files. Also note we are working with strings here and not `RelativePath`, making the conversion a little bit harder.

What would be a good use-case for having relative paths here?

Comment (danyf90) [reply projects/docs/docs/commands/graph.md:139]: Going from relative to absolute should be just a matter of prepending a fixed path, so I don't think it's a big issue.
Any logic you would do on the files, I expect you do it from the project root, so you should be able to use relative paths directly.

I don't have a strong use-case for relative paths, but even being able to compare generated JSONs on different machines could be a nice to have.
I think we should go for relative by default everywhere, unless we have a strong reason to use absolute.

Makes sense? 🚀

Comment (fortmarek) [reply projects/docs/docs/commands/graph.md:139]: I'm fine with that, once we start using this in tasks, we should provide a utility that will convert a string of relative path to a string of absolute path, so it's as easy-as-possible for users to use.

File: Sources/TuistKit/Services/GraphService.swift (hunk 20 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:80]: Yeah that sounds good to me.

File: Sources/TuistKit/Services/GraphService.swift (hunk 21 above)
Comment (neakor) [reply Sources/TuistKit/Services/GraphService.swift:78]: I'm going to resolve this here since as you mentioned, this is not relevant to this PR.

File: projects/docs/docs/commands/graph.md (hunk 23 above)
Comment (neakor) [reply projects/docs/docs/commands/graph.md:21]: Major doesn't necessarily mean breaking though. I'm not sure if we should break the schema in minor versions.

File: Tests/TuistKitTests/Services/GraphServiceTests.swift (hunk 22 above)
Comment (neakor) [reply Tests/TuistKitTests/Services/GraphServiceTests.swift:69]: What do you mean? It is checking the results by deserialize it. Marek has suggested to do this.

File: projects/docs/docs/commands/graph.md (hunk 24 above)
Comment (neakor) [reply projects/docs/docs/commands/graph.md:139]: For our use case we actually must have absolute paths. Using relative paths would make our life a lot more painful. Using absolute paths is also consistent with SPM's output. I feel pretty strongly that we should stick with absolute paths here.

Comment (fortmarek) [reply projects/docs/docs/commands/graph.md:139]: Fair! The current format is labeled as unstable, so if we find relative paths would actually be better, I'd revert it. But so far it seems that absolute paths are a better option here.

File: projects/docs/docs/commands/graph.md (hunk 23 above)
Comment (fortmarek) [reply projects/docs/docs/commands/graph.md:21]: > I'm not sure if we should break the schema in minor versions.

I think we should put out a warning this might break even in minor versions. This enables to iterate fast on this part of the codebase without scaring users by putting out too many major versions.

Comment (neakor) [reply projects/docs/docs/commands/graph.md:21]: Ok sounds good!

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistGraph/Output/PackageOutput.swift [hunk 25]
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+/// The structure defining the output schema of an Swift package.
```
[the remark this answers is not in the document]
Comment (neakor) [reply Sources/TuistGraph/Output/PackageOutput.swift:3]: Oops! Thank you.

File: Sources/TuistGraph/Output/ProjectOutput.swift [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+        schemes: [SchemeOutput] = [SchemeOutput]()) {
+        self.name = name
+        self.path = path
+        self.packages = packages
+        self.targets = targets
+        self.schemes = schemes
+    }
+    
+    /// Factory function to convert an internal graph project to the output type.
+    public static func from(_ project: Project) -> ProjectOutput {
```
[the remark this answers is not in the document]
Comment (neakor) [reply Sources/TuistGraph/Output/ProjectOutput.swift:35]: Yeah sounds good.

**Review by fortmarek — APPROVED** [review]
LGTM, thanks a lot @neakor 👏

**Comment by neakor** [comment]
Thank you team for helping review and pushing this through!
